OWL on FHIR

joeflack4 opened the pull request:
Major updates
```
OWL Conversion
- Add: Script to do the conversion
- Add: requirements*.txt files for managing dependencies.
- Add: Docs to README.md
- Add: robot.jar, a current, temporary dependency for the conversion.
```

joeflack4 committed:
OWL conversion script

joeflack4 committed:
OWL Conversion
- Add: Script to do the conversion
- Add: requirements*.txt files for managing dependencies.
- Add: Docs to README.md
- Add: robot.jar, a current, temporary dependency for the conversion.

Misc
- .gitignore: Added '/output/' and `bin/env.sh`. Small amount of re-organization.
- Update: bin/start-ts.sh: Now copies over `env.sh.example` to `env.sh`
- Update: `bin/env.sh` is now `bin/env.sh.example`

Docs/README.md
- Add: OWL conversion info
- Add: Information about ports

joeflack4 commented on bin/convert-owl.py:
```diff
@@ -0,0 +1,113 @@
+):
+    """Run conversion"""
+    # Download if necessary
+    # TODO: determine if is URL or path. if URL, download and set new input_path
```
Still needs to be done:
> `    # TODO: determine if is URL or path. if URL, download and set new input_path`

joeflack4 replied:
done

joeflack4 commented on bin/convert-owl.py:
```diff
@@ -0,0 +1,113 @@
+    obograph = owl_to_obograph(input_path)
+
+    # TODO: save obograph file
+    obograph_path = ''
```
Still needs to be done:
> `    # TODO: save obograph file`

I forgot I would need to do this step. When I was writing my own converter, I just kept the Obograph in memory without saving to disk. But for the current OAK implementation, I need to save it and re-read.

joeflack4 replied:
done

joeflack4 commented on bin/convert-owl.py:
```diff
@@ -0,0 +1,113 @@
+        '-o', '--out-dir', required=False, default=OUTDIR, help='The directory where results should be saved.')
+    parser.add_argument(
+        '-n', '--out-filename', required=False, help='Filename for the primary file converted, e.g. CodeSystem.')
+    # TODO: When https://github.com/INCATools/ontology-access-kit/pull/374 is merged, can add include_all_predicates
```
Left out `include_all_predicates` flag for now, because the latest release of `oaklib` doesn't have this yet. Chris merged my PR, but still need a new release.

joeflack4 commented on bin/convert-owl.py:
```diff
@@ -0,0 +1,113 @@
+        '-n', '--out-filename', required=False, help='Filename for the primary file converted, e.g. CodeSystem.')
+    # TODO: When https://github.com/INCATools/ontology-access-kit/pull/374 is merged, can add include_all_predicates
+    # parser.add_argument(
+    #     '-p', '--predicates-is-a-only', action='store_true', default=False,
```
For parameterization of this conversion function, I felt it made sense to invert the `--include-all-predicates` flag to `--predicates-is-a-only'` for our purposes, since I think that will work better for our use case. But I can just leave the flag param the same way as we're using in OAK (`--include-all-predicates`).

For now, still waiting anyway on a new OAK release to enable this.

joeflack4 commented on bin/convert-owl.py:
```diff
@@ -0,0 +1,113 @@
+    print()
+
+
+def cli():
```
FYI: I made a CLI anyway, which you don't have to use. This is just easier for me to work with and useful until you get to that.

joeflack4 commented on bin/convert-owl.py:
```diff
@@ -0,0 +1,113 @@
+    converter.dump(gd, out_path)
+
+
+def owl_to_fhir(
```
@ShahimEssaid This is the function you will want to use with your CLI. Still needs some tweaking before this work can be merged, though.

joeflack4 replied:
done

joeflack4 reviewed (commented):
@ShahimEssaid Here's the initial draft. Should be able to finish up by Friday.

I also created this repository that we talked about: https://github.com/HOT-Ecosystem/owl-on-fhir-content/

After I'm done writing this script, I'll convert the rest of those files and upload to that repo.

joeflack4 commented on README.md:
```diff
@@ -2,8 +2,16 @@
 See [this doc](https://docs.google.com/document/d/1J7exF73Z0Z8W-SCYRrQ5qNQVMZTkIH0c4rf-sgjDBfk) for work in progress documentation and notes.
+
+## Workflows
+### Converting OWL
```
@ShahimEssaid I know you plan on not running this directly, but importing the function and using your own CLI, but for now, here are some docs.

joeflack4 replied:
removed in favor of docs PR and [google doc](https://docs.google.com/document/d/1J7exF73Z0Z8W-SCYRrQ5qNQVMZTkIH0c4rf-sgjDBfk/edit#)

joeflack4 commented on bin/convert_owl.py:
```diff
@@ -0,0 +1,130 @@
+PROJECT_DIR = os.path.join(BIN_DIR, '..')
+CACHE_DIR = os.path.join(PROJECT_DIR, 'cache')
+OUTDIR = os.path.join(PROJECT_DIR, 'output')
+ROBOT_PATH = os.path.join(BIN_DIR, 'robot')
```
@ShahimEssaid I had to add `robot.jar` to do the obographs conversion for now.

joeflack4 commented on bin/convert_owl.py:
```diff
@@ -0,0 +1,130 @@
+
+
+# Functions
+def download(url: str, path: str, download_if_cached=True):
```
FYI: the download function

joeflack4 commented on bin/convert_owl.py:
```diff
@@ -0,0 +1,130 @@
+            f.write(response.content)
+
+
+def owl_to_obograph(inpath: str, outpath: str):
```
FYI: how OWL is converted to obograph

joeflack4 commented on bin/convert_owl.py:
```diff
@@ -0,0 +1,130 @@
+
+
+# todo: When https://github.com/INCATools/ontology-access-kit/pull/374 is merged, can add include_all_predicates
+def obograph_to_fhir(inpath: str, out_dir: str, out_filename: str = None, include_all_predicates=False):
```
FYI: how obograph is converted to fhir

joeflack4 commented on bin/convert_owl.py:
```diff
@@ -0,0 +1,130 @@
+    converter.dump(gd, out_path)
+
+
+def owl_to_fhir(
```
@ShahimEssaid This is the function you would want to import and tie to your CLI.

ShahimEssaid commented:
@joeflack4 I will review and use this asap but the general idea is that the python code/script will go into the "setup-servers" repository, wrapped in a Click cli command, etc. The bash scripts under /bin will invoke that functionality from "setup-servers" as needed.  The robot Jar will be downloaded on demand by the "setup-servers" command and not be committed in the repo (just like how the Maven distribution is downloaded and used on demand)

joeflack4 commented:
I don't disagree w/ that approach. I don't see any `setup-servers` repo though. Is this a directory that will be created by the scripts that build and run the server?

I don't mind `robot.jar` being downloaded, but it would be nice to have a way to convert OWL to FHIR in a standalone way. That's what I was trying to do [here](https://github.com/joeflack4/owl-on-fhir/), out of the context/coupling of anything like OAK, the terminology server code, etc. And I could pre-convert stuff to simply be downloaded so that it doesn't need to be built by the server. Even though I'm griping, I realize there's a lot of different ways to do this, and I can be flexible.

joeflack4 commented on README.md:
```diff
@@ -2,8 +2,27 @@
 
 See [this doc](https://docs.google.com/document/d/1J7exF73Z0Z8W-SCYRrQ5qNQVMZTkIH0c4rf-sgjDBfk) for work in progress documentation and notes.
+
+## Prerequisites
```
Added some prerequisites. I didn't know what the minimum Python version should be so I just put 3.8

joeflack4 replied:
Removed and added to docs PR. i see its in [google doc](https://docs.google.com/document/d/1J7exF73Z0Z8W-SCYRrQ5qNQVMZTkIH0c4rf-sgjDBfk/edit#) as well

joeflack4 commented on README.md:
```diff
@@ -2,8 +2,27 @@
+3. Docker
+
+## Server management
+### Starting the server
```
New section. This is where I added how to access the server when it starts. Figured this should be front and center,  not just in the google doc.

joeflack4 replied:
removed and added to docs pr

joeflack4 commented on bin/env.sh:
```diff
@@ -4,4 +4,5 @@
-SS_VENV=
\ No newline at end of file
+SS_VENV=
+SS_VENV=.venv
```
@ShahimEssaid I Added this file to the gitignore but it has still been committed previously still needs to be removed manually. I can do that if you want but, but figured you might want to do.

joeflack4 commented on bin/start-ts.sh:
```diff
@@ -17,6 +17,7 @@ done
 DIR="$(cd -P "$(dirname "$SOURCE")" && pwd)"
 SS_HOME=$(realpath "${DIR}/..")
 cd "$SS_HOME"
+cp bin/env.sh.example bin/env.sh
```
The pattern I was talking about on Slack to separate the file that is needed and should not be changed (`.example`) from the file that will actually be edited (but we don't want git to see as changed).

joeflack4 replied:
removed as Shahim already implemented in `main` and ive rebased

joeflack4 commented on requirements.txt:
```diff
@@ -0,0 +1,141 @@
+aiohttp==3.8.3
```
The Python dependencies that need to be installed for the OWL conversion, transitive dependencies fully resolved. See `requirements-unlocked.txt` for the base requirements (only 3 entries).

joeflack4 replied:
I'm not sure where you are storing other Python requirements that need to be installed, but we should probably combine those requirements with these.

joeflack4 commented on bin/convert_owl.py:
```diff
@@ -0,0 +1,142 @@
+CACHE_DIR = os.path.join(PROJECT_DIR, 'cache')
+OUTDIR = os.path.join(PROJECT_DIR, 'output')
+ROBOT_PATH = os.path.join(BIN_DIR, 'robot')
+FAVORITE_ONTOLOGY_URLS = {
```
@ShahimEssaid Thought it would be prudent to store these URLs somewhere.

joeflack4 committed:
Rolled back some changes not related to OWL-on-FHIR feature. Moved OWL-on-FHIR docs to Google Doc

joeflack4 committed:
OWL on FHIR
- Dependencies: Upgraded OAK
- Add: SemanticSQL intermediary format functionality.
- Add: CLI: --intermediary-type
- Add: CLI: --use-cached-intermediaries
- Update: CLI: Changed --predicates-is-a-only to --include-all-predicates
- Bugfix: Obographs: There is an issue w/ the data format itself, as well as how robot and LinkML are handling it, where in some cases you can have an unexpected `null` that results in an unhanled exception in OAK. Removed instances of these `null`s.

Misc
- Update: .gitignore: Added entry for OAK/semsql

joeflack4 commented on bin/convert_owl.py:
```diff
@@ -0,0 +1,188 @@
+
+
+# TODO:
+def semsql_to_fhir(inpath: str, out_dir: str, out_filename: str = None, include_all_predicates=False) -> str:
```
`semsql` to FHIR converter

joeflack4 commented on bin/convert_owl.py:
```diff
@@ -0,0 +1,188 @@
+    return result
+
+
+def owl_to_semsql(input_path: str) -> str:
```
OWL to `semsql` converter

joeflack4 committed:
OWL on FHIR
- Update: Removed patching around genusId logical axioms with nulls inside them, as this has now been addressed in OAK.

joeflack4 committed:
OWL on FHIR
- Add: RXNORM: It is a special case, so some things had to be done specifically for it.
- Update: robot.jar upgraded from 1.8.1 to 1.9.1. This includes a fix with 'null' appearing in output, which previously was a bug. OAK handles this now, but still good to fix the problem upstream.

joeflack4 commented:
@ShahimEssaid I think I forgot to undraft this before. This has been ready to merge.

joeflack4 commented on bin/convert_owl_ncbo2owl.pl:
```diff
@@ -0,0 +1,8 @@
+#!/usr/bin/perl -np
```
This is for some special modifications that need to be made to Bioportal's RxNorm.

joeflack4 committed:
- Add: URI for Sequence Ontology (SO)

joeflack4 committed:
- Bugfix: Added a fix in Obographs (filed an issue as well) where there were some concepts which should have appeared in 'nodes' but did not

joeflack4 committed:
- Update: Temporarily deactivated prior bugfix on missing concepts because it is unclear whether they should indeed be there and it's just an OAK issue that is preventing conversion (we can upload to FHIR server), or if they should not be there
- Update: Added a fix for robot 1.9.1 which generates a correct Obograph, but OAK cannot yet convert it. This is temporary until that fix and removes some extra properties

joeflack4 committed:
- Delete: Patch to Obographs converter which removed the 'meta' property. OAK latest release fixes
- Add: Functionality 'favorites' for converting everything at once with a config.

joeflack4 commented on bin/convert_owl.py:
```diff
@@ -0,0 +1,390 @@
+    return os.path.join(out_dir, out_filename)
+
+
+def _run_favorites(
```
@ShahimEssaid This way of converting 'favorites' is the biggest part of this PRthat I wonder if you might have issue with. I can do this differently, but also I can remove it altogether and move it to my own personal script outside of VCS or in my own separate repo.

joeflack4 committed:
OWL on FHIR
- Update: Decreased verbosity of passing arguments using favorites()

joeflack4 committed:
OWL on FHIR
- Add: param: --native-uri-stems: A comma-separated list of URI stems that will be used to determine whether a concept is native to the CodeSystem. As of 2023-01-15, there is still a bug in the Obographs spec and/or `robot` where certain nodes are not being converted. This converter adds back the nodes, but to know which ones belong to the CodeSystem itself and are not foreign concepts, this parameter is necessary. OAK also makes use of this parameter. See also: https://github.com/geneontology/obographs/issues/90
- Add: params: --dev-oak-path and --dev-oak-interpreter-path: If you want to use a local development version of OAK, specify these paths.

joeflack4 commented:
@ShahimEssaid It looks like [OWL-on-FHIR](https://github.com/HOT-Ecosystem/owl-on-fhir/) has been successfully refactored out, so I'm going to close this PR now.

It actually was working correctly by the time we all met last Monday the 20th. The next step is for me to merge my OAK PR, so until then, OWL-on-FHIR won't work as a package and will need to be cloned and used with a local clone of OAK. I'm going to try to get that down now during my current work session.
